### ISSUE TITLE
CustomCVForce using RMSDForce CV incurs significant performance impact even with small number of restrained atoms

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#2156]
We noticed that adding a `CustomCVForce` using an `RMSDForce` seems to incur a significant performance penalty even when a very small number of atoms are restrained.

Here's the output of the attached simple demonstration script when run on a GTX-1080Ti using a recent OpenMM 7.3 `dev` build (`38ae256`):
```
                                      without RMSD CustomCVForce :             CUDA :  172.021 ns/day
                                      without RMSD CustomCVForce :           OpenCL :  128.745 ns/day
Restraining 13 protein atoms and 29 ligand atoms (50608 atoms total)
                                      with    RMSD CustomCVForce :             CUDA :  108.584 ns/day
                                      with    RMSD CustomCVForce :           OpenCL :   87.666 ns/day
```
Note that only 29 atoms / 50608 are restrained.

Is there an easy route to improving the performance here via further optimization?

Alternatively, if the optimal alignment code is the culprit but incapable of being optimized, would it be possible to allow RMSDForce to optionally *not* perform the rotation prior to RMSD computation as an optional argument?

[rmsd-restraint.zip](https://github.com/pandegroup/openmm/files/2299860/rmsd-restraint.zip)

Update: A repeated run shows slightly faster performance, but still demonstrates a significant performance hit:
```
[chodera@lilac:rmsd-restraint]$ python time-rmsd-restraint.py
                                      without RMSD CustomCVForce :             CUDA :  178.856 ns/day
                                      without RMSD CustomCVForce :           OpenCL :  132.685 ns/day
Restraining 13 protein atoms and 29 ligand atoms (50608 atoms total)
                                      with    RMSD CustomCVForce :             CUDA :  138.957 ns/day
                                      with    RMSD CustomCVForce :           OpenCL :  121.040 ns/day
```

**Comment by peastman** [comment]
It's a mix of both.  CustomCVForce needs to copy information from the main context to its internal context, compute the forces on that, wait for them to finish, and then copy them back.  That takes time and the main context is blocked for part of it.  In principle I should be able to improve it by moving most of that work to a separate thread, but when I tried doing that it ended up being slower.  I'm not sure why.

RMSDForce is a bit slow because it does the SVD on the CPU, which requires round trip communication.  In principle it might be possible to speed it up by doing that on the GPU.  I haven't yet tried doing that.

**Comment by jchodera** [comment]
> It's a mix of both. CustomCVForce needs to copy information from the main context to its internal context, compute the forces on that, wait for them to finish, and then copy them back. That takes time and the main context is blocked for part of it. In principle I should be able to improve it by moving most of that work to a separate thread, but when I tried doing that it ended up being slower. I'm not sure why.

Interesting! I wonder if this is true on all hardware. If not, maybe this could be a context parameter or environment variable setting? It *does* seem like that would be faster computed in parallel.

> RMSDForce is a bit slow because it does the SVD on the CPU, which requires round trip communication. In principle it might be possible to speed it up by doing that on the GPU. I haven't yet tried doing that.

I think the NVIDIA-provided [`cuSOLVER` library](https://docs.nvidia.com/cuda/cusolver/index.html) provides a [SVD solver](https://docs.nvidia.com/cuda/cusolver/index.html#cuds-lt-t-gt-gesvd).

**Comment by jchodera** [comment]
> CustomCVForce needs to copy information from the main context to its internal context, compute the forces on that, wait for them to finish, and then copy them back. 

Also, in this case, we only need a small subset of the coordinates to compute the RMSDForce. I wonder if there's some way to take advantage of the fact that most forces one would want to use as CVs only involve a small subset of particles.

**Comment by andrrizzi** [comment]
Following this, I've ran a benchmark of our restraint forces. We recently converted them to use a `CustomCVForce` to be able to extract the restraint distances easily from the simulation. Here are the results.

```
OLD RESTRAINTS
--------------
without     any restraint :             CUDA :  183.518 ns/day
with   Harmonic restraint :             CUDA :  180.821 ns/day
with FlatBottom restraint :             CUDA :  182.485 ns/day

CVForce RESTRAINTS
------------------
without     any restraint :             CUDA :  184.622 ns/day
with   Harmonic restraint :             CUDA :  124.776 ns/day
with FlatBottom restraint :             CUDA :  145.675 ns/day
```

Do you see an optimization of `CVCustomForce` that would help reduce the gap between the two implementations? Or an alternative way to extract, for example, the harmonic restraint distance that doesn't involve a `CustomCVForce`? If this is not possible, we'll probably have to provide two implementations, one using the `CVCustomForce` and one without.

**Comment by peastman** [comment]
What sort of restraints are these?  Can you describe how the old implementation worked, and how you implemented them with CustomCVForce?

**Comment by jchodera** [comment]
@andrrizzi : Can you post a ZIP archive of your code?

@peastman: All of the restraints we use only make use of a small subset of atomic positions (including the RMSD restraints). I wonder if this could be exploited to speed up `CustomCVForce`, where when we initialize it, we tell it what particles to copy over/back (or perhaps it could figure this out automatically).

**Comment by andrrizzi** [comment]
Of course, sorry. The old implementation of the harmonic restraint is a simply
```python
CustomCentroidBondForce('(K/2)*distance(g1,g2)^2')
# configure g1 and g2 to be a set of protein and ligand receptors.
``` 
The new implementation is something like
```python
cv_force = CustomCVForce('(K/2)*restraint_distance^2')
distance_force = CustomCentroidBondForce(2, 'distance(g1,g2)')
# configure groups and bond of CustomCentroidBondForce
cv_force.addCollectiveVariable('restraint_distance', distance_force)
```

**Comment by andrrizzi** [comment]
Here are all the files: [benchmark-restraints.zip](https://github.com/choderalab/yank/files/2310934/benchmark-restraints.zip)

**Comment by peastman** [comment]
CustomCVForce seems unnecessarily complex for that case.  Why not just put the CustomCentroidBondForce in its own force group so you query it individually?

> when we initialize it, we tell it what particles to copy over/back (or perhaps it could figure this out automatically).

That wouldn't make a difference.  All the copying is done directly on the GPU, so it takes very little time.

**Comment by jchodera** [comment]
`CustomCVForce` is currently the *only* way to use the RMSD in a restraint expression. Also, we found that we need to use multiple restraints at a time and store them each iteration of simulation, and bundling all of these possible restraints in a `CustomCVForce` is a simple way to keep all of these together and generally iterate over them to store them each iteration.

We *could* rely on different force groups for this with some significant re-architecting, but we'd still need a `CustomRMSDForce` that supports using the computed RMSD in an energy expression in order to even do that.

**Comment by andrrizzi** [comment]
> We switched to the `CustomCVForce` for our restraints so we could quickly query distances, angles, and dihedrals

As Levi said, this is the main reason of the new implementation.

> Why not just put the `CustomCentroidBondForce` in its own force group so you query it individually?

Do you mean inferring the `distance(g1,g2)` from the energy of its force group and the energy function? On the top of my head, that solution would be hard to generalize for any restraint we (or the library's users) will want to add, but maybe there's something I'm missing. Is there an alternative way to read directly the value of `distance(g1, g2)` from the `Context`?

**Comment by Lnaden** [comment]
> Is there an alternative way to read directly the value of distance(g1, g2) from the Context?

Since we know what atoms are in `g1` and `g2`, we can compute the distances from the coordinates at run time. I don't think there is a built-in way to get an arbitrary `Custom*Force` function though (e.g. `distance(X,Y)`, `angle(X,Y,Z)`, etc.)

**Comment by andrrizzi** [comment]
> we can compute the distances from the coordinates at run time.

I'd like to avoid that as, in the general case, I think that will require imaging the system.

**Comment by peastman** [comment]
I was referring to the code you posted above.  There are lots of cases where CustomCVForce is required, but your example didn't seem to be one of them.

**Comment by andrrizzi** [comment]
What I meant is that the `CustomCVForce` implementation is required if we want to read the value of `distance(g1,g2)` from the `Context`. At least, I'm not familiar with other ways of obtaining the same result.

**Comment by jchodera** [comment]
Here are some updated timings for all restraints with and without `CustomCVForce` on a GTX-1080Ti:
```
[chodera@lilac:benchmark-restraints]$ python time-restraints.py
Restraining 13 protein atoms and 29 ligand atoms (50608 atoms total)
                                              without restraints :             CUDA :  185.625 ns/day
                                                       RMSDForce :             CUDA :  152.593 ns/day :   17.795 % slowdown
                                      RMSDForce in CustomCVForce :             CUDA :  103.138 ns/day :   44.438 % slowdown
                                              harmonic restraint :             CUDA :  183.268 ns/day :    1.270 % slowdown
                             harmonic restraint in CustomCVForce :             CUDA :  156.691 ns/day :   15.588 % slowdown
                                           flat-bottom restraint :             CUDA :  182.858 ns/day :    1.491 % slowdown
                          flat-bottom restraint in CustomCVForce :             CUDA :  127.639 ns/day :   31.238 % slowdown
```
[benchmark-restraints.zip](https://github.com/pandegroup/openmm/files/2364595/benchmark-restraints.zip)

**Comment by peastman** [comment]
I don't think it's really about the number of particles.  All the copying of positions and forces is done on the GPU, so that should be very fast.  The problem is that it introduces a bubble into the pipeline.  I still hope to find a way to improve that, but my attempts so far haven't helped.

**Comment by schuhmc** [comment]
Is there any intention to pick this up again? I was just testing an RMSD restraint on a small system (<800 particles) in implicit solvent, and the performance drop is still very significant.

On a 4060-Ti I am getting about 770 ns/day with RMSD restraints, and 1930 ns/day without - that's a 60% slowdown. Similar story on a L40s.

**Comment by peastman** [comment]
This issue was closed many years ago.  If it's still a problem for you, can you open a new issue?  I don't know if there's much that can be done, especially for such a tiny system, but we can look into it.

**Comment by schuhmc** [comment]
Thanks, I'll look into how things look like in a larger system and then consider opening a new issue. Looking at other RMSD restraint implementations I found this comment in Colvars https://github.com/Colvars/colvars/pull/541#issuecomment-1640293733 and it seems like their overhead is around 6%, but that is supposedly also for a much larger system.
